File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlSourceOperations.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.source.mysql;

import static com.mysql.cj.MysqlType.BIGINT;
import static com.mysql.cj.MysqlType.BIGINT_UNSIGNED;
import static com.mysql.cj.MysqlType.DATE;
import static com.mysql.cj.MysqlType.DATETIME;
import static com.mysql.cj.MysqlType.DECIMAL;
import static com.mysql.cj.MysqlType.DECIMAL_UNSIGNED;
import static com.mysql.cj.MysqlType.DOUBLE;
import static com.mysql.cj.MysqlType.DOUBLE_UNSIGNED;
import static com.mysql.cj.MysqlType.FLOAT;
import static com.mysql.cj.MysqlType.FLOAT_UNSIGNED;
import static com.mysql.cj.MysqlType.INT;
import static com.mysql.cj.MysqlType.INT_UNSIGNED;
import static com.mysql.cj.MysqlType.LONGTEXT;
import static com.mysql.cj.MysqlType.MEDIUMINT;
import static com.mysql.cj.MysqlType.MEDIUMINT_UNSIGNED;
import static com.mysql.cj.MysqlType.MEDIUMTEXT;
import static com.mysql.cj.MysqlType.SMALLINT;
import static com.mysql.cj.MysqlType.SMALLINT_UNSIGNED;
import static com.mysql.cj.MysqlType.TEXT;
import static com.mysql.cj.MysqlType.TIME;
import static com.mysql.cj.MysqlType.TIMESTAMP;
import static com.mysql.cj.MysqlType.TINYINT;
import static com.mysql.cj.MysqlType.TINYINT_UNSIGNED;
import static com.mysql.cj.MysqlType.TINYTEXT;
import static com.mysql.cj.MysqlType.VARCHAR;
import static com.mysql.cj.MysqlType.YEAR;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_COLUMN_NAME;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_COLUMN_SIZE;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_COLUMN_TYPE;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_COLUMN_TYPE_NAME;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_DECIMAL_DIGITS;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_SCHEMA_NAME;
import static io.airbyte.db.jdbc.JdbcConstants.INTERNAL_TABLE_NAME;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.NullNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.mysql.cj.MysqlType;
import com.mysql.cj.jdbc.result.ResultSetMetaData;
import com.mysql.cj.result.Field;
import io.airbyte.db.SourceOperations;
import io.airbyte.db.jdbc.AbstractJdbcCompatibleSourceOperations;
import io.airbyte.protocol.models.JsonSchemaType;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.format.DateTimeParseException;
import java.util.Set;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MySqlSourceOperations extends AbstractJdbcCompatibleSourceOperations<MysqlType> implements SourceOperations<ResultSet, MysqlType> {

  private static final Logger LOGGER = LoggerFactory.getLogger(MySqlSourceOperations.class);
  private static final Set<MysqlType> ALLOWED_CURSOR_TYPES = Set.of(TINYINT, TINYINT_UNSIGNED, SMALLINT,
      SMALLINT_UNSIGNED, MEDIUMINT, MEDIUMINT_UNSIGNED, INT, INT_UNSIGNED, BIGINT, BIGINT_UNSIGNED,
      FLOAT, FLOAT_UNSIGNED, DOUBLE, DOUBLE_UNSIGNED, DECIMAL, DECIMAL_UNSIGNED, DATE, DATETIME, TIMESTAMP,
      TIME, YEAR, VARCHAR, TINYTEXT, TEXT, MEDIUMTEXT, LONGTEXT);

  /**
   * @param colIndex 1-based column index.
   */
  @Override
  public void copyToJsonField(final ResultSet resultSet, final int colIndex, final ObjectNode json) throws SQLException {
    final ResultSetMetaData metaData = (ResultSetMetaData) resultSet.getMetaData();
    final Field field = metaData.getFields()[colIndex - 1];
    final String columnName = field.getName();
    final MysqlType columnType = field.getMysqlType();

    // https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html
    switch (columnType) {
      case BIT -> {
        if (field.getLength() == 1L) {
          // BIT(1) is boolean
          putBoolean(json, columnName, resultSet, colIndex);
        } else {
          putBinary(json, columnName, resultSet, colIndex);
        }
      }
      case BOOLEAN -> putBoolean(json, columnName, resultSet, colIndex);
      case TINYINT -> {
        if (field.getLength() == 1L) {
          // TINYINT(1) is boolean
          putBoolean(json, columnName, resultSet, colIndex);
        } else {
          putShortInt(json, columnName, resultSet, colIndex);
        }
      }
      case TINYINT_UNSIGNED, YEAR -> putShortInt(json, columnName, resultSet, colIndex);
      case SMALLINT, SMALLINT_UNSIGNED, MEDIUMINT, MEDIUMINT_UNSIGNED -> putInteger(json, columnName, resultSet, colIndex);
      case INT, INT_UNSIGNED -> {
        if (field.isUnsigned()) {
          putBigInt(json, columnName, resultSet, colIndex);
        } else {
          putInteger(json, columnName, resultSet, colIndex);
        }
      }
      case BIGINT, BIGINT_UNSIGNED -> putBigInt(json, columnName, resultSet, colIndex);
      case FLOAT, FLOAT_UNSIGNED -> putFloat(json, columnName, resultSet, colIndex);
      case DOUBLE, DOUBLE_UNSIGNED -> putDouble(json, columnName, resultSet, colIndex);
      case DECIMAL, DECIMAL_UNSIGNED -> {
        if (field.getDecimals() == 0) {
          putBigInt(json, columnName, resultSet, colIndex);
        } else {
          putBigDecimal(json, columnName, resultSet, colIndex);
        }
      }
      case DATE -> putDate(json, columnName, resultSet, colIndex);
      case DATETIME -> putTimestamp(json, columnName, resultSet, colIndex);
      case TIMESTAMP -> putTimestampWithTimezone(json, columnName, resultSet, colIndex);
      case TIME -> putTime(json, columnName, resultSet, colIndex);
      case CHAR, VARCHAR -> putString(json, columnName, resultSet, colIndex);
      case TINYBLOB, BLOB, MEDIUMBLOB, LONGBLOB, BINARY, VARBINARY, GEOMETRY -> putBinary(json, columnName, resultSet, colIndex);
      case TINYTEXT, TEXT, MEDIUMTEXT, LONGTEXT, JSON, ENUM, SET -> putString(json, columnName, resultSet, colIndex);
      case NULL -> json.set(columnName, NullNode.instance);
      default -> putDefault(json, columnName, resultSet, colIndex);
    }
  }

  /**
   * MySQL boolean is equivalent to tinyint(1).
   */
  @Override
  protected void putBoolean(final ObjectNode node, final String columnName, final ResultSet resultSet, final int index) throws SQLException {
    node.put(columnName, resultSet.getInt(index) > 0);
  }

  @Override
  public void setCursorField(final PreparedStatement preparedStatement,
                             final int parameterIndex,
                             final MysqlType cursorFieldType,
                             final String value)
      throws SQLException {
    switch (cursorFieldType) {
      case BIT -> setBit(preparedStatement, parameterIndex, value);
      case BOOLEAN -> setBoolean(preparedStatement, parameterIndex, value);
      case YEAR, TINYINT, TINYINT_UNSIGNED, SMALLINT, SMALLINT_UNSIGNED, MEDIUMINT, MEDIUMINT_UNSIGNED -> setInteger(preparedStatement,
          parameterIndex,
          value);
      case INT, INT_UNSIGNED, BIGINT, BIGINT_UNSIGNED -> setBigInteger(preparedStatement, parameterIndex, value);
      case FLOAT, FLOAT_UNSIGNED, DOUBLE, DOUBLE_UNSIGNED -> setDouble(preparedStatement, parameterIndex, value);
      case DECIMAL, DECIMAL_UNSIGNED -> setDecimal(preparedStatement, parameterIndex, value);
      case DATE -> setDate(preparedStatement, parameterIndex, value);
      case DATETIME -> setTimestamp(preparedStatement, parameterIndex, value);
      case TIMESTAMP -> setTimestampWithTimezone(preparedStatement, parameterIndex, value);
      case TIME -> setTime(preparedStatement, parameterIndex, value);
      case CHAR, VARCHAR, TINYTEXT, TEXT, MEDIUMTEXT, LONGTEXT, ENUM, SET -> setString(preparedStatement, parameterIndex, value);
      case TINYBLOB, BLOB, MEDIUMBLOB, LONGBLOB, BINARY, VARBINARY -> setBinary(preparedStatement, parameterIndex, value);
      // since cursor are expected to be comparable, handle cursor typing strictly and error on
      // unrecognized types
      default -> throw new IllegalArgumentException(String.format("%s cannot be used as a cursor.", cursorFieldType));
    }
  }

  @Override
  public MysqlType getDatabaseFieldType(final JsonNode field) {
    try {
      // MysqlType#getByName can handle the full MySQL type name
      // e.g. MEDIUMINT UNSIGNED
      final MysqlType literalType = MysqlType.getByName(field.get(INTERNAL_COLUMN_TYPE_NAME).asText());
      final int columnSize = field.get(INTERNAL_COLUMN_SIZE).asInt();
      switch (literalType) {
        // BIT(1) and TINYINT(1) are interpreted as boolean
        case BIT, TINYINT -> {
          if (columnSize == 1) {
            return MysqlType.BOOLEAN;
          }
        }
        case YEAR -> {
          return SMALLINT;
        }
        // When CHAR[N] and VARCHAR[N] columns have binary character set, the returned
        // types are BINARY[N] and VARBINARY[N], respectively. So we don't need to
        // convert them here. This is verified in MySqlSourceDatatypeTest.
        case DECIMAL -> {
          if (field.get(INTERNAL_DECIMAL_DIGITS) != null && field.get(INTERNAL_DECIMAL_DIGITS).asInt() == 0) {
            return BIGINT;
          }
        }
        case DECIMAL_UNSIGNED -> {
          if (field.get(INTERNAL_DECIMAL_DIGITS) != null && field.get(INTERNAL_DECIMAL_DIGITS).asInt() == 0) {
            return BIGINT_UNSIGNED;
          }
        }
      }
      return literalType;
    } catch (final IllegalArgumentException ex) {
      LOGGER.warn(String.format("Could not convert column: %s from table: %s.%s with type: %s (type name: %s). Casting to VARCHAR.",
          field.get(INTERNAL_COLUMN_NAME),
          field.get(INTERNAL_SCHEMA_NAME),
          field.get(INTERNAL_TABLE_NAME),
          field.get(INTERNAL_COLUMN_TYPE),
          field.get(INTERNAL_COLUMN_TYPE_NAME)));
      return MysqlType.VARCHAR;
    }
  }

  @Override
  public boolean isCursorType(final MysqlType type) {
    return ALLOWED_CURSOR_TYPES.contains(type);
  }

  @Override
  public JsonSchemaType getAirbyteType(final MysqlType mysqlType) {
    return switch (mysqlType) {
      case
      // TINYINT(1) is boolean, but it should have been converted to MysqlType.BOOLEAN in {@link
      // getFieldType}
      TINYINT, TINYINT_UNSIGNED, SMALLINT, SMALLINT_UNSIGNED, INT, MEDIUMINT, MEDIUMINT_UNSIGNED, INT_UNSIGNED, BIGINT, BIGINT_UNSIGNED -> JsonSchemaType.INTEGER;
      case FLOAT, FLOAT_UNSIGNED, DOUBLE, DOUBLE_UNSIGNED, DECIMAL, DECIMAL_UNSIGNED -> JsonSchemaType.NUMBER;
      case BOOLEAN -> JsonSchemaType.BOOLEAN;
      case NULL -> JsonSchemaType.NULL;
      // BIT(1) is boolean, but it should have been converted to MysqlType.BOOLEAN in {@link getFieldType}
      case BIT, TINYBLOB, BLOB, MEDIUMBLOB, LONGBLOB, BINARY, VARBINARY, GEOMETRY -> JsonSchemaType.STRING_BASE_64;
      case TIME -> JsonSchemaType.STRING_TIME_WITHOUT_TIMEZONE;
      case DATETIME -> JsonSchemaType.STRING_TIMESTAMP_WITHOUT_TIMEZONE;
      case TIMESTAMP -> JsonSchemaType.STRING_TIMESTAMP_WITH_TIMEZONE;
      case DATE -> JsonSchemaType.STRING_DATE;
      default -> JsonSchemaType.STRING;
    };
  }

  @Override
  protected void setDate(final PreparedStatement preparedStatement, final int parameterIndex, final String value) throws SQLException {
    try {
      preparedStatement.setObject(parameterIndex, LocalDate.parse(value));
    } catch (final DateTimeParseException e) {
      // This is just for backward compatibility for connectors created on versions before PR
      // https://github.com/airbytehq/airbyte/pull/15504
      LOGGER.warn("Exception occurred while trying to parse value for date column the new way, trying the old way", e);
      super.setDate(preparedStatement, parameterIndex, value);
    }
  }

  @Override
  protected void setTimestamp(final PreparedStatement preparedStatement, final int parameterIndex, final String value) throws SQLException {
    try {
      preparedStatement.setObject(parameterIndex, LocalDateTime.parse(value));
    } catch (final DateTimeParseException e) {
      // This is just for backward compatibility for connectors created on versions before PR
      // https://github.com/airbytehq/airbyte/pull/15504
      LOGGER.warn("Exception occurred while trying to parse value for datetime column the new way, trying the old way", e);
      preparedStatement.setObject(parameterIndex, OffsetDateTime.parse(value));
    }
  }

  private void setTimestampWithTimezone(final PreparedStatement preparedStatement, final int parameterIndex, final String value) throws SQLException {
    try {
      preparedStatement.setObject(parameterIndex, OffsetDateTime.parse(value));
    } catch (final DateTimeParseException e) {
      // This is just for backward compatibility for connectors created on versions before PR
      // https://github.com/airbytehq/airbyte/pull/15504
      LOGGER.warn("Exception occurred while trying to parse value for timestamp column the new way, trying the old way", e);
      preparedStatement.setObject(parameterIndex, LocalDateTime.parse(value));
    }
  }

}

File: airbyte-integrations/connectors/source-mysql/src/main/kotlin/io/airbyte/integrations/source/mysql/MySqlSourceOperations.kt
/* Copyright (c) 2024 Airbyte, Inc., all rights reserved. */
package io.airbyte.integrations.source.mysql

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.mysql.cj.MysqlType
import io.airbyte.cdk.command.OpaqueStateValue
import io.airbyte.cdk.discover.CdcIntegerMetaFieldType
import io.airbyte.cdk.discover.CdcOffsetDateTimeMetaFieldType
import io.airbyte.cdk.discover.CdcStringMetaFieldType
import io.airbyte.cdk.discover.CommonMetaField
import io.airbyte.cdk.discover.Field
import io.airbyte.cdk.discover.FieldType
import io.airbyte.cdk.discover.JdbcAirbyteStreamFactory
import io.airbyte.cdk.discover.JdbcMetadataQuerier
import io.airbyte.cdk.discover.MetaField
import io.airbyte.cdk.discover.SystemType
import io.airbyte.cdk.jdbc.BigDecimalFieldType
import io.airbyte.cdk.jdbc.BigIntegerFieldType
import io.airbyte.cdk.jdbc.BinaryStreamFieldType
import io.airbyte.cdk.jdbc.BooleanFieldType
import io.airbyte.cdk.jdbc.BytesFieldType
import io.airbyte.cdk.jdbc.DoubleFieldType
import io.airbyte.cdk.jdbc.FloatFieldType
import io.airbyte.cdk.jdbc.IntFieldType
import io.airbyte.cdk.jdbc.JdbcFieldType
import io.airbyte.cdk.jdbc.LocalDateFieldType
import io.airbyte.cdk.jdbc.LocalDateTimeFieldType
import io.airbyte.cdk.jdbc.LocalTimeFieldType
import io.airbyte.cdk.jdbc.LongFieldType
import io.airbyte.cdk.jdbc.LosslessJdbcFieldType
import io.airbyte.cdk.jdbc.NullFieldType
import io.airbyte.cdk.jdbc.OffsetDateTimeFieldType
import io.airbyte.cdk.jdbc.PokemonFieldType
import io.airbyte.cdk.jdbc.ShortFieldType
import io.airbyte.cdk.jdbc.StringFieldType
import io.airbyte.cdk.read.And
import io.airbyte.cdk.read.Equal
import io.airbyte.cdk.read.From
import io.airbyte.cdk.read.FromNode
import io.airbyte.cdk.read.FromSample
import io.airbyte.cdk.read.Greater
import io.airbyte.cdk.read.GreaterOrEqual
import io.airbyte.cdk.read.Lesser
import io.airbyte.cdk.read.LesserOrEqual
import io.airbyte.cdk.read.Limit
import io.airbyte.cdk.read.LimitNode
import io.airbyte.cdk.read.NoFrom
import io.airbyte.cdk.read.NoLimit
import io.airbyte.cdk.read.NoOrderBy
import io.airbyte.cdk.read.NoWhere
import io.airbyte.cdk.read.Or
import io.airbyte.cdk.read.OrderBy
import io.airbyte.cdk.read.OrderByNode
import io.airbyte.cdk.read.SelectColumnMaxValue
import io.airbyte.cdk.read.SelectColumns
import io.airbyte.cdk.read.SelectNode
import io.airbyte.cdk.read.SelectQuery
import io.airbyte.cdk.read.SelectQueryGenerator
import io.airbyte.cdk.read.SelectQuerySpec
import io.airbyte.cdk.read.Stream
import io.airbyte.cdk.read.Where
import io.airbyte.cdk.read.WhereClauseLeafNode
import io.airbyte.cdk.read.WhereClauseNode
import io.airbyte.cdk.read.WhereNode
import io.airbyte.cdk.read.cdc.DebeziumState
import io.airbyte.cdk.util.Jsons
import io.micronaut.context.annotation.Primary
import jakarta.inject.Singleton
import java.time.OffsetDateTime

@Singleton
@Primary
class MySqlSourceOperations :
    JdbcMetadataQuerier.FieldTypeMapper, SelectQueryGenerator, JdbcAirbyteStreamFactory {

    override val globalCursor: MetaField = MySqlSourceCdcMetaFields.CDC_CURSOR

    override val globalMetaFields: Set<MetaField> =
        setOf(
            MySqlSourceCdcMetaFields.CDC_CURSOR,
            CommonMetaField.CDC_UPDATED_AT,
            CommonMetaField.CDC_DELETED_AT,
            MySqlSourceCdcMetaFields.CDC_LOG_FILE,
            MySqlSourceCdcMetaFields.CDC_LOG_POS
        )

    override fun decorateRecordData(
        timestamp: OffsetDateTime,
        globalStateValue: OpaqueStateValue?,
        stream: Stream,
        recordData: ObjectNode
    ) {
        recordData.set<JsonNode>(
            CommonMetaField.CDC_UPDATED_AT.id,
            CdcOffsetDateTimeMetaFieldType.jsonEncoder.encode(timestamp),
        )
        recordData.set<JsonNode>(
            MySqlSourceCdcMetaFields.CDC_LOG_POS.id,
            CdcIntegerMetaFieldType.jsonEncoder.encode(0),
        )
        if (globalStateValue == null) {
            return
        }
        val debeziumState: DebeziumState =
            MySqlSourceDebeziumOperations.deserializeDebeziumState(globalStateValue)
        val position: MySqlSourceCdcPosition =
            MySqlSourceDebeziumOperations.position(debeziumState.offset)
        recordData.set<JsonNode>(
            MySqlSourceCdcMetaFields.CDC_LOG_FILE.id,
            CdcStringMetaFieldType.jsonEncoder.encode(position.fileName),
        )
        recordData.set<JsonNode>(
            MySqlSourceCdcMetaFields.CDC_LOG_POS.id,
            CdcIntegerMetaFieldType.jsonEncoder.encode(position.position),
        )
    }

    override fun toFieldType(c: JdbcMetadataQuerier.ColumnMetadata): FieldType =
        when (val type = c.type) {
            is SystemType -> leafType(type)
            else -> PokemonFieldType
        }

    private fun leafType(type: SystemType): JdbcFieldType<*> {
        return when (MysqlType.getByName(type.typeName)) {
            MysqlType.BIT -> if (type.precision == 1) BooleanFieldType else BytesFieldType
            MysqlType.BOOLEAN -> BooleanFieldType
            MysqlType.TINYINT,
            MysqlType.TINYINT_UNSIGNED,
            MysqlType.YEAR -> ShortFieldType
            MysqlType.SMALLINT,
            MysqlType.SMALLINT_UNSIGNED,
            MysqlType.MEDIUMINT,
            MysqlType.MEDIUMINT_UNSIGNED,
            MysqlType.INT -> IntFieldType
            MysqlType.INT_UNSIGNED,
            MysqlType.BIGINT -> LongFieldType
            MysqlType.BIGINT_UNSIGNED -> BigIntegerFieldType
            MysqlType.FLOAT,
            MysqlType.FLOAT_UNSIGNED,
            MysqlType.DOUBLE,
            MysqlType.DOUBLE_UNSIGNED -> {
                if ((type.precision ?: 0) <= 23) FloatFieldType else DoubleFieldType
            }
            MysqlType.DECIMAL,
            MysqlType.DECIMAL_UNSIGNED -> {
                if (type.scale == 0) BigIntegerFieldType else BigDecimalFieldType
            }
            MysqlType.DATE -> LocalDateFieldType
            MysqlType.DATETIME -> LocalDateTimeFieldType
            MysqlType.TIMESTAMP -> OffsetDateTimeFieldType
            MysqlType.TIME -> LocalTimeFieldType
            MysqlType.CHAR,
            MysqlType.VARCHAR,
            MysqlType.TINYTEXT,
            MysqlType.TEXT,
            MysqlType.MEDIUMTEXT,
            MysqlType.LONGTEXT,
            MysqlType.ENUM,
            MysqlType.SET -> StringFieldType
            MysqlType.JSON -> StringFieldType // TODO: replace this with JsonStringFieldType
            MysqlType.TINYBLOB,
            MysqlType.BLOB,
            MysqlType.MEDIUMBLOB,
            MysqlType.LONGBLOB,
            MysqlType.BINARY,
            MysqlType.VARBINARY,
            MysqlType.GEOMETRY -> BinaryStreamFieldType
            MysqlType.NULL -> NullFieldType
            MysqlType.VECTOR,
            MysqlType.UNKNOWN,
            null -> PokemonFieldType
        }
    }

    override fun generate(ast: SelectQuerySpec): SelectQuery =
        SelectQuery(ast.sql(), ast.select.columns, ast.bindings())

    fun SelectQuerySpec.sql(): String {
        val components: List<String> = listOf(select.sql(), from.sql(), where.sql(), orderBy.sql())
        val sqlWithoutLimit: String = components.filter { it.isNotBlank() }.joinToString(" ")
        val rownumClause: String =
            when (limit) {
                NoLimit -> return sqlWithoutLimit
                Limit(0) -> "LIMIT 0"
                is Limit -> "LIMIT ?"
            }
        return "$sqlWithoutLimit $rownumClause"
    }

    fun SelectNode.sql(): String =
        "SELECT " +
            when (this) {
                is SelectColumns -> columns.joinToString(", ") { it.sql() }
                is SelectColumnMaxValue -> "MAX(${column.sql()})"
            }

    fun Field.sql(): String = "`$id`"

    fun FromNode.sql(): String =
        when (this) {
            NoFrom -> ""
            is From -> if (this.namespace == null) "FROM `$name`" else "FROM `$namespace`.`$name`"
            is FromSample -> {
                val from: String = From(name, namespace).sql()
                // On a table that is very big we limit sampling to no less than 0.05%
                // chance of a row getting picked. This comes at a price of bias to the beginning
                // of table on very large tables ( > 100s million of rows)
                val greatestRate: String = 0.00005.toString()
                // We only do a full count in case information schema contains no row count.
                // This is the case for views.
                val fullCount = "SELECT COUNT(*) FROM `$namespace`.`$name`"
                // Quick approximation to "select count(*) from table" which doesn't require
                // full table scan. However, note this could give delayed summary info about a table
                // and thus a new table could be treated as empty despite we recently added rows.
                // To prevent that from happening and resulted for skipping the table altogether,
                // the minimum count is set to 10.
                val quickCount =
                    "SELECT GREATEST(10, COALESCE(table_rows, ($fullCount))) FROM information_schema.tables WHERE table_schema = '$namespace' AND table_name = '$name'"
                val greatest = "GREATEST($greatestRate, $sampleSize / ($quickCount))"
                // Rand returns a value between 0 and 1
                val where = "WHERE RAND() < $greatest "
                "$from $where"
            }
        }

    fun WhereNode.sql(): String =
        when (this) {
            NoWhere -> ""
            is Where -> "WHERE ${clause.sql()}"
        }

    fun WhereClauseNode.sql(): String =
        when (this) {
            is And -> conj.joinToString(") AND (", "(", ")") { it.sql() }
            is Or -> disj.joinToString(") OR (", "(", ")") { it.sql() }
            is Equal -> "${column.sql()} = ?"
            is Greater -> "${column.sql()} > ?"
            is GreaterOrEqual -> "${column.sql()} >= ?"
            is LesserOrEqual -> "${column.sql()} <= ?"
            is Lesser -> "${column.sql()} < ?"
        }

    fun OrderByNode.sql(): String =
        when (this) {
            NoOrderBy -> ""
            is OrderBy -> "ORDER BY " + columns.joinToString(", ") { it.sql() }
        }

    fun SelectQuerySpec.bindings(): List<SelectQuery.Binding> = where.bindings() + limit.bindings()

    fun WhereNode.bindings(): List<SelectQuery.Binding> =
        when (this) {
            is NoWhere -> listOf()
            is Where -> clause.bindings()
        }

    fun WhereClauseNode.bindings(): List<SelectQuery.Binding> =
        when (this) {
            is And -> conj.flatMap { it.bindings() }
            is Or -> disj.flatMap { it.bindings() }
            is WhereClauseLeafNode -> {
                val type = column.type as LosslessJdbcFieldType<*, *>
                listOf(SelectQuery.Binding(bindingValue, type))
            }
        }

    fun LimitNode.bindings(): List<SelectQuery.Binding> =
        when (this) {
            NoLimit,
            Limit(0) -> listOf()
            is Limit -> listOf(SelectQuery.Binding(Jsons.numberNode(n), LongFieldType))
        }
}
